### DISCUSSION TITLE
Isn't it nice if enemies AI didn't know exact amount of my troops in the castle?

### DISCUSSION THREAD

**Discussion by vtarelkin** [ihhub/fheroes2#9266] [Ideas]
Somehow computer enemies knows what is inside my castle and decide to attack (or not) basing on that information. Is the same logic implemented in original Heroes? It seems a little bit unfair

**Comment by oleg-derevenetz** [comment]
> Somehow computer enemies knows what is inside my castle and decide to attack (or not) basing on that information.

Yes, AI knows the army composition of visible enemies (including army composition in garrisons), otherwise, it would not be able to perform estimations and make adequate decisions.

> Is the same logic implemented in original Heroes?

In the original HoMM2 AI implemented differently, but yes, it also knows the enemy army composition, consider:

https://github.com/user-attachments/assets/efd181f8-7c96-427c-b081-ec8c0f5fc299

> It seems a little bit unfair

Consider this a kind of response to the save/load abuse practiced by humans :)

**Reply by vtarelkin** [reply]
Maybe add more human-like switchable AI? Like, I mean, when the enemie is near the garrison he make a decision whether to attack or not based on some prediction not on the count of the creatures. 
Also there is an annoying bug: when my hero is inside the garrison the enemy won't attack and when I step out it runs over the whole map to attack. I can return and enemy starts to roaming around and when I left again he again runs to my castle and so on and on...

**Reply by oleg-derevenetz** [reply]
You can always offer a pull request with the specific implementation of your ideas. It will be reviewed, evaluated, tested, and if it is qualitative and improves some aspect of AI behavior without regressions, it will be merged.

> Also there is an annoying bug: when my hero is inside the garrison the enemy won't attack and when I step out it runs over the whole map to attack.

Castles are fairly high-priority targets, even at a considerable distance, so the AI will try to capture them in the absence of any really high-priority targets.
